### ISSUE TITLE
Custom metadata support

### ISSUE THREAD

**Issue by pmonks** [chaoss/grimoirelab#37]
When there's additional custom metadata available for a data source that's being ingested into GrimoireLabs / Bitergia, I want to be able to capture and index that metadata into the system, so that I can query and report on it the same as the "system" metadata that GrimoireLabs / Bitergia already captures.

Examples of such metadata include:

* Project lifecycle information (e.g. current state and lifecycle stage transition dates)
* Project contribution date (which is _not_ always the same as repository creation date or first commit date)
* Member / customer flag
* Type of organisation (e.g. FinServ vs FinTech vs other)

**Comment by acs** [comment]
@pmonks do you have a specific list of metadata you already have? Just to four listed above? We are working in a new tool for Project Description (Bestiary: https://github.com/acs/bestiary) and this ticket could be related to it.

**Comment by pmonks** [comment]
@acs those four, plus "Program" (a string value) are our current metadata requirements, however I think it's inevitable that we'll identify additional metadata needs going forward.

I think it's also important to highlight that these four are quite specific to our business problem and are unlikely to be applicable to other organisations, so a general purpose metadata mechanism that allowed each user to define & populate their own custom metadata model would be highly preferable.

**Comment by pmonks** [comment]
I'd encourage you to review the content management literature for ideas on how to properly support this requirement.  Custom metadata has been a central part of that industry segment for more than 3 decades, so there's a lot of prior art to crib from.

tl;dr - supporting custom metadata models is easy; supporting the evolution of custom metadata models over time is where things get complex and can't really be retrofitted.

**Comment by acs** [comment]
@pmonks thanks for your suggestions. The initial approach probably will be to define the `metadata` as a JSON (python dictionary) and store it as a string serialized in a field in the project data model. And once we have real use cases of the metadata, we can design a more semantic approach to the metadata management.

**Comment by pmonks** [comment]
@acs that sounds like a good start, but recall that metadata isn't just for projects - it may also be for users, affiliations (companies), data sources, etc.

**Comment by mindthegab** [comment]
Hey @acs and team - is there any chance we can get an update on this one? Is it actually planned for a certain release or timeframe? 

The inflexibility of bitergia model is becoming quite of a blocker as our foundation grows, which at this point does not solve important business problems for us (project lifecycle, project categorization, organization type, etc).

Thanks in advance for your attention on this!

**Comment by acs** [comment]
@mindthegab first of all, I am sorry for the lack of a timely response to this ticket.

Recovering the initial requirements from @pmonks, I will focus in some specific use cases, for supporting the projects metadata, we have already a meta field in the projects.json file, which can be used, but later, we need to use this data during the index generation so the data appears,

For adding metadata to member and organizations, we need to do it in SortingHat, in order to have all the identities data in a unique source.

But this is not a general solution, I know. For working in a general solution, we will need a common data schema for all the items in GrimoireLab, so we can add metadata to them. And right not, this common data schema does not exists.

What GrimoireLab has is that all items in the platform are described in JSON format, so during the generation of the data for the indexes, the "metadata source" could be queried to add additional info to the JSON items.

Right now we are redefining the enrichment processes so it is a good moment to analyze this requirement. But we can not provide a commitment for such a general feature now.

If we focus in projects or identities metadata, we can analyze them to provide you a release or timeframe estimation.

Do you think @mindthegab and @pmonks that we can focus in some initial use cases (having in mind during the design that the idea is to extend it to the rest of GriomoireLab)?

**Comment by pmonks** [comment]
@acs custom metadata for just projects would be a great start - that's where most of our requirements are.

**Comment by acs** [comment]
> @acs custom metadata for just projects would be a great start - that's where most of our requirements are.

Great. Let's focus on that then. The implementation is pretty direct So we just need to define it correctly. Let's start adding a new field to the "meta" field in projects.json called "ecosystem". Could you add it to your projects.json in SymphonyOSS, @pmonks ? After that, we will add support so this field appears in all your enriched indexes just in the same way that project field now appear. 

So for example:

```
  "WG-API" : {
    "confluence" : [ "WGA" ],
    "mbox"       : [ "wg-api /home/bitergia/mboxes/barnowl_symphony_wg-api" ],
    "meta"       : { "title" : "WG-API", "ecosystem": "GlobalECO"}
  }
```

for all the items enriched related to the WG_API in confluence and mbox, there will be a field called `metatada_custom_ecosystem=GlobalECO`.

Does it satisfy your needs?

**Comment by pmonks** [comment]
Yes, although it would ideal if we could shorten the `metadata_custom_` prefix to make querying in the Kibana UI easier.  How about using `mc_` or (even better) `cm_` instead?

**Comment by acs** [comment]
> Yes, although it would ideal if we could shorten the metadata_custom_ prefix to make querying in the Kibana UI easier. How about using mc_ or (even better) cm_ instead?
> 

Yes, it makes sense. I like the "cm_" prefix. Let's go with it in the first iteration.

> Also, can we also start putting in the other custom project related fields I listed above? Meaning, is this already an extensible, general purpose way to add any fields we like?

The ideas as that all fields that appears inside the 'meta' dictionary will appear in the items, yes.

> And if so, are there any restrictions we should keep in mind (e.g. total number of custom fields, use of non-string data types, etc.)?

If the custom field values are valid JSON values, I think it should be ok. So for example, you can have as values strings, numbers, lists, dicts ... and these will appear in the enriched items. Let's start with string, numbers and lists and check that the chain support them. My bet is that there will be no problems.

**Comment by pmonks** [comment]
Perfect.  For right now I think we only have string values anyway, though I could imagine adding a date field before too long (which may be tricky, given JSON / Javascript appalling lack of support for date/time/TZ data types).

**Comment by acs** [comment]
> Perfect. For right now I think we only have string values anyway, though I could imagine adding a date field before too long (which may be tricky, given the appalling lack of support for date/time/TZ data types in JSON / Javascript).

JSON is which fixes some requirements here. In Python we just get the JSON and convert it to a python dict. So if you can add it to JSON, we can add it to the enriched items :+1:

**Comment by acs** [comment]
All is implemented now. In the `projects.json` file you can add metadata to a project which will be shown in all its items.

For example:

```
....
  "FDX Program" : {
    "meta"       : { "title"   : "FDX Program",
                         "program" : "FDX" },
...
```

So closing this item.
